src/matcher/regex.js

```diff
@@ -14,7 +25,151 @@ import { noOpObj } from '@keg-hub/jsutils'
 export const matchRegex = (step, text) => {
   const match = text.match(new RegExp(step.match))
 
-  // If there is a match, get all matching items except for the first one
   // Which is the original string
-  return match ? { step, match: match.slice(1, match.length) } : noOpObj
+  return match
+    ? { step, match: match.slice(1, match.length).filter(Boolean) }
+    : noOpObj
+}
+
+/**
+ * Converts an optional expression into regex
+ * @param {string} optional 
+ * @return {string} regex for an optional cucumber-expression 
+ * @example
+ * toAlternateRegex('test(s)')
+ * result: '(test|tests)'
+ */
+export const toAlternateRegex = optional => {
+  const split = optional.split(/(\(|\))/)
+
+  const [ start, , middle, , end ] = split
+
+  // no words outside of optional boundary
+  if (start === '' && end === '')
+    return optional + '?'
+  else if (start === '')
+    return `(${middle}|${middle}${end})`
+  else if (end === '')
+    return `(${start}|${start}${middle})`
+  else
+    return `(${start}${end}|${start}${middle}${end})`
+}
+
+/**
+ * Gets the full text around an optional
+ * @param {Array<string>} match result of optional regex match
+ */
+const getFullOptionalText = (match) => {
+  const text = match.input
+  const precedingWord = getWordEndingAt(text, match.index)
+  return precedingWord + match[0]
+}
+
+/**
+ * Helper for `getParamRegex` to get the optional types regex
+ * @param {Array<string>} match result of optional regex match
+ * @return {string} the correct regex source text for an step definition
+ * optional part. This regex will be used for matching values in the
+ * feature step text.
+ */
+const getOptionalRegex = match => {
+  const optionalText = getFullOptionalText(match)
+  return toAlternateRegex(optionalText)
+}
+
+/**
+ * Returns regex for a given parameter type
+ * @param {string} type - cucumber-expression parameter type: float, int, word, or string
+ * @return {string} regex source text
+ */
+export const getParamRegex = type => {
```

Comment:
Instead of adding this helper here, the regex for each type
  * Should be added to it's corresponding transformer in the paramTypes file
    * There's already a field for it in the `typeModel` each param is based off of

Then instead of calling this helper on line 80 of `expression.js`
  * Just use the already found transformer, with `transformer.regexp`

You can then import the `getParamTypes` into the `regex.js` file from `paramTypes.js`
And in the `getParamRegex`, use the `transformer.regexp` property there based on the passed in type
Looks like this => 
```js
// Add this to the top of regex.js
 import { getParamTypes } from './paramTypes'
 
const getMatchRegex = (type, match) => {
  const [ val, paramType ] = match
  // Get the types at the time it was called, so it's up to date
  const paramTypes = getParamTypes()

  switch(type) {
    case 'parameter':
      // Create the new RegExp from the paramTypes regexp property
      return new RegExp(paramTypes[paramType].regexp)

     // ...rest of the codes
```

Reply:
Nice I like this. Updated